src/rpc.rs

```diff
@@ -1,103 +1,378 @@
+			},
+			_ => (),
+		}
+
+		info!("Connection established to the node: {full_node_ws} <{version}>");
+		let node = Node {
+			host: full_node_ws.to_owned(),
+			system_version,
+			spec_version: client.runtime_version().spec_version,
+			genesis_hash,
+		};
 
-pub async fn get_valset_by_hash(
-	client: &avail::Client,
-	hash: H256,
-) -> Result<Vec<ed25519::Public>> {
-	let grandpa_valset = client
-		.runtime_api()
-		.at(hash)
-		.call_raw::<Vec<(ed25519::Public, u64)>>("GrandpaApi_grandpa_authorities", None)
+		Ok((client, node))
+	}
+
+	async fn connect_to_available_rpc(
+		full_nodes: &[String],
+		expected_version: &ExpectedVersion<'static>,
+		expected_genesis_hash: Option<H256>,
+	) -> Result<(avail::Client, Node)> {
+		full_nodes
+			.iter()
+			.map(|address| {
+				Self::connect(address, expected_version, expected_genesis_hash)
+					.inspect_err(move |error| warn!(address, %error, "Skipping connection"))
+			})
+			.collect::<futures::stream::FuturesUnordered<_>>()
+			.skip_while(|res| futures::future::ready(res.is_err()))
+			.map(Result::unwrap)
+			.next()
+			.await
+			.context("Failed to connect to a working node")
+	}
+
+	/// Shuffles full nodes to randomize access,
+	/// and pushes last full node to the end of a list
+	/// so we can try it if connection to other node fails
+	fn shuffle_full_nodes(full_nodes: &mut Vec<String>, last_full_node: Option<&String>) {
+		let old_len = full_nodes.len();
+		full_nodes.retain(|node| Some(node) != last_full_node);
+		full_nodes.shuffle(&mut thread_rng());
+
+		// Pushing last full node to the end of a list, if it's only one left to try
+		if let (Some(node), true) = (last_full_node, old_len != full_nodes.len()) {
+			full_nodes.push(node.clone());
+		}
+	}
+
+	pub async fn new(
+		nodes: Vec<String>,
+		expected_version: ExpectedVersion<'static>,
+		db: Option<Arc<DB>>,
+	) -> Result<Self> {
+		let expected_genesis_hash = if let Some(db) = &db {
+			crate::data::get_genesis_hash(db.clone())?
+		} else {
+			None
+		};
+		let (client, node) =
+			Self::connect_to_available_rpc(&nodes, &expected_version, expected_genesis_hash)
+				.await?;
+
+		info!(?node.genesis_hash);
+		if let Some(db) = &db {
+			crate::data::store_last_full_node_ws_in_db(db.clone(), node.host.clone())?;
+			if expected_genesis_hash.is_none() {
+				info!("No genesis hash is found in the db, storing the new hash now.");
+				crate::data::store_genesis_hash(db.clone(), node.genesis_hash)?;
+			}
+		}
+
+		let backoff = backoff::ExponentialBackoffBuilder::new()
+			.with_max_elapsed_time(Some(std::time::Duration::from_secs(20)))
+			.build();
+		Ok(Self {
+			client: Arc::new(RwLock::new((client, node))),
+			nodes,
+			expected_version,
+			db,
+			backoff,
+		})
+	}
+
+	async fn current_client(&self) -> avail::Client {
+		self.client.read().await.0.clone()
+	}
+
+	pub async fn current_node(&self) -> Node {
+		self.client.read().await.1.clone()
+	}
+
+	pub async fn with_client<F, Fut, T>(&self, mut f: F) -> Result<T>
```

Comment:
It would be nice if this function is rewritten for better comprehension, it has lots of moving parts, it's difficult to understand. Some comments would also not be bad. Also, this is central to the entire PR, this function needs to be in the best possible state. Let's spend some time improving this.

Reply:
Can you suggest how it can be improved? I can add comments describing what I do here, but apart from that I don't know how to improve it :man_shrugging:

Reply:
Added comments